questionpy_server/cache.py

```diff
@@ -0,0 +1,152 @@
+        """
+        Creates a path of the given filename. It is not mandatory that the filename exists.
+
+        :param filename: the filename
+        :return: path of the file
+        """
+
+        return self.directory / (filename + self._extension)
+
+    def contains(self, key: str) -> bool:
+        """
+        Checks if the file exists in cache and places it to the end ensuring that it is the most recent accessed file.
+        """
+
+        if key not in self._files:
+            return False
+
+        file = self._files[key]
+
+        if not file.path.is_file():
+            # File was deleted from filesystem - update internal dict.
+            self._total_bytes -= file.size
+            del self._files[key]
+            return False
+
+        self._files.move_to_end(key)
+        return True
+
+    def _get_file(self, key: str) -> File:
+        if not self.contains(key):
+            raise FileNotFoundError
+
+        return self._files[key]
+
+    def get(self, key: str) -> Path:
+        return self._get_file(key).path
+
+    def remove(self, key: str) -> None:
+        file = self._get_file(key)
+
+        file.path.unlink(missing_ok=True)
+        self._total_bytes -= file.size
+
+        del self._files[key]
+
+    def clear(self) -> None:
+        for file in self._files.values():
+            file.path.unlink(missing_ok=True)
+
+        self._total_bytes = 0
+        self._files.clear()
+
+    def put(self, key: str, value: bytes) -> Path:
+        """
+        Only accepts `bytes` objects as values; raises a `TypeError` otherwise.
+        If the length/size of the provided `value` exceeds `.max_bytes` a `SizeError` is raised.
+        The internal `._total_bytes` attribute is updated.
+        If the key existed before and just the value is replaced, the item is treated as most recently accessed and
+        thus moved to the end of the internal linked list.
+        If after adding the item `._total_bytes` exceeds `.max_bytes`, items are deleted in order from least to most
+        recently accessed until the total size (in bytes) is in line with the specified maximum.
+        """
+
+        if not isinstance(value, bytes):
+            raise TypeError("Not a bytes object:", repr(value))
+
+        size = len(value)
+        if size > self.max_bytes:
+            # If we allowed this, the loop at the end would remove all items from the dictionary,
+            # so we raise an error to allow exceptions for this case.
+            raise SizeError(f"Item itself exceeds maximum allowed size of {self.max_bytes} bytes",
+                            max_size=self.max_bytes, actual_size=size)
+
+        # Update `_total_bytes` depending on whether the key existed already or not.
+        self._total_bytes += size - self._get_file(key).size if key in self._files else size
+
+        # Save the bytes on filesystem.
+        path = self._path(key)
+        assert size == path.open('wb').write(value)
```

Comment:
1. Es fehlt hier ein `close()` call. Oder noch besser mit `with`.

   > Calling f.write() without using the with keyword or calling f.close() might result in the arguments of f.write() not being completely written to the disk, even if the program exits successfully.
   https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files
2. Nebeneffekte in `assert`-Statements halte ich nicht für super gut. Zwar werden sie in Python zwar standardmäßig ausgeführt, aber mit `-O` oder der Umgebungsvariable `PYTHONOPTIMIZE` werden sie ausgeschaltet.

```suggestion
        with path.open("wb") as file:
            written_bytes = file.write(value)
            assert size == written_bytes
```

Reply:
1. Da hast du absolut Recht. Ich wollte ursprünglich die Funktion `path.write_bytes(value)` verwenden, welche die Datei automatisch öffnet und schließt - werde es anpassen.
2. Ich habe das `assert`-Statement mit einer Exception ersetzt.